### ISSUE TITLE
Margin for the cover page

### ISSUE THREAD

**Issue by andkorsh** [wkhtmltopdf/wkhtmltopdf#3635]
Hi wkhtml team!

First of all, thank you for the great product! I don't know how can I thank you more - this product is really great! Thanks again - you're doing the great job.

My question actually has been asked multiple times at different forums but unfortunately there were no answer for that. And I think it is not possible within the current wkhtml API - probably you guys can advice if there is an ability to at least patch/build the custom version. We really need that..

The thing is that we don't want margin-top and margin-bottom to be applied to the COVER page. That's fine that it is applied to the TOC and to the main pages, but we don't really want it to be applied to the cover page. We'd like to have an ability either to apply different margins for a cover page or not to apply margins for cover page at all.

If it is possible to patch the wkhtml library and build the custom version for us - I'd appreciate your help guys.

The version I'm running is "wkhtmltopdf 0.12.3.2 (with patched qt)".

Thanks,
Andrew K.

**Comment by andkorsh** [comment]
OK, 1 day and done :)
https://github.com/andkorsh/wkhtmltopdf/commit/c09862d6878fcf2f5cd406b05651c5bb8420f375

I can make a pull request if needed, please let me know.

Thanks again!

**Comment by bitbonk** [comment]
I want this feature1

**Comment by NiR-** [comment]
@andkorsh You should submit a PR for that :)

**Comment by dogabner** [comment]
I want this, too!

**Comment by andkorsh** [comment]
@bitbonk @NiR- @eliiik @katesowles here you go https://github.com/wkhtmltopdf/wkhtmltopdf/pull/3692 :)

**Comment by hakimio** [comment]
Does anyone have pre-built binaries for Linux 64bit with this patch?

**Comment by guicara** [comment]
Thank you for the PR, it will be really great to have this feature

However, it will be really nice to set different margins for any page! Let's take for example a magazine with a cover page, a table of contents and articles in many categories (ordered by categories). I would like to have a kind of "cover" page per category.

> - Cover page *
> - Table of contents
> - Cover page for the category "A" *
> - 1st article of the category A
> - 2nd article of the cateogry A
> - Cover page for the category "B" *
> - 1st article of the category B
> - 2nd article of the cateogry B
> 
> * (without any margin to display a full background image)

**Comment by hakimio** [comment]
@guicara good luck implementing your suggestion ;)

**Comment by guicara** [comment]
@hakimio Is it really different from the PR request of @andkorsh?

**Comment by hakimio** [comment]
@guicara He just removed top and bottom margins from the cover page. Nothing more.

**Comment by PhilterPaper** [comment]
To remove or change margins on an arbitrary interior page would probably better be done with embedded directives (special HTML comments), although I suppose with sufficient pain and suffering you could find a way to specify margins from the command line for a given absolute page number. While you're at it, the ability to control most command line settings (not just margins) would probably be helpful.

With the current code + above PR, if you want "interior covers", perhaps you could build a series of documents, each with its own cover, and then use a separate utility to merge all the PDFs into one document. You would probably want to control the printed page numbering when you do that, so that it's continuous and not starting over each time.

**Comment by guicara** [comment]
Thank you for your reply @PhilterPaper.

> To remove or change margins on an arbitrary interior page would probably better be done with embedded directives (special HTML comments)

👍 

> (...) if you want "interior covers", perhaps you could build a series of documents, each with its own cover, and then use a separate utility to merge all the PDFs into one document

This is indeed maybe the best solution in the meantime. I have end up doing that on my project : I use the PHP wrapper [phpwkhtmltopdf](https://github.com/mikehaertl/phpwkhtmltopdf) to generate multiple temporary PDF files, then [php-pdftk](https://github.com/mikehaertl/php-pdftk) to merge all the files.

Two major drawbacks:

- like you said, it is a pain to handle page numbering
- the `wkhtmltopdf` table of contents does not work well with multiple pages and merged files

This issue is out of my scope... (C++).
It would be greatly appreciated if someone have time to implementing that :-)

**Comment by Arvin-He** [comment]
It will be very nice to have this feature.

**Comment by stuartcusackie** [comment]
I agree. This would be good for my needs too. I'm aiming for a coloured cover page with no margins, and the rest of the document to have margins.

I tried to merge two separate pdfs but it's a pain and it doesn't work well. Also you lose your page numbering.

**Comment by hakimio** [comment]
@jfcartier Andrew made a simple proof of concept which you can find [here](https://github.com/wkhtmltopdf/wkhtmltopdf/pull/3692). If you really want this feature, you'll have to put some effort to make it ready for merging as at this point it's nowhere near that stage.

**Comment by darmawan01** [comment]
is there any solution for this ? to set different margin each page

**Comment by hakimio** [comment]
@darmawan01 No, you can track progress [here](https://github.com/wkhtmltopdf/wkhtmltopdf/pull/3692).

**Comment by enisn** [comment]
Any update?

**Comment by hakimio** [comment]
@enisn No, you will have to work on PR #3692 yourself if you want this functionality. See my comments ([1](https://github.com/wkhtmltopdf/wkhtmltopdf/pull/3692#issuecomment-354750967) & [2](https://github.com/wkhtmltopdf/wkhtmltopdf/pull/3692#issuecomment-354955879)) about what needs to be fixed and possible solution to some of the issues.

**Comment by t0byman** [comment]
Need this!

**Comment by cristinabardas** [comment]
Need this feature as well!
